src/ndarray/ndarray.cc

```diff
@@ -124,24 +124,50 @@ void BinaryOp(const NDArray &lhs,
   // redirect everything to mshadow operations
   switch (lhs.ctx().dev_mask()) {
     case cpu::kDevMask: {
-      Engine::Get()->PushSync([lhs, rhs, ret](RunContext ctx) {
-          ret.CheckAndAlloc();
-          TBlob tmp = ret.data();
-          ndarray::Eval<cpu, OP>(lhs.data(), rhs.data(), &tmp, ctx);
-        }, lhs.ctx(), const_vars, {ret.var()},
-        FnProperty::kNormal, 0, PROFILER_MESSAGE_FUNCNAME);
+      if (lhs.storage_type() == kDefaultStorage && rhs.storage_type() == kDefaultStorage) {
+        Engine::Get()->PushSync([lhs, rhs, ret](RunContext ctx) {
+            ret.CheckAndAlloc();
+            TBlob tmp = ret.data();
+            ndarray::Eval<cpu, OP>(lhs.data(), rhs.data(), &tmp, ctx);
+          }, lhs.ctx(), const_vars, {ret.var()},
+          FnProperty::kNormal, 0, PROFILER_MESSAGE_FUNCNAME);
+      } else if (lhs.storage_type() == rhs.storage_type()) {
+        Engine::Get()->PushSync([lhs, rhs, ret](RunContext ctx) {
+            ret.CheckAndAlloc();
+            // variable captured by value is const
+            // need to use tmp as a delegate
+            NDArray tmp = ret;
+            ndarray::Eval<cpu, OP>(lhs, rhs, &tmp, ctx);
+          }, lhs.ctx(), const_vars, {ret.var()},
+          FnProperty::kNormal, 0, PROFILER_MESSAGE_FUNCNAME);
+      } else {
```

Comment:
Maybe it's okay to skip conversion and fail here, since operator overload is only used at backend. And i think it's not frequently used..?